lgc/util/CacheHashNoteUtils.cpp

```diff
@@ -0,0 +1,298 @@
+    initialOffset = sectionAndNewOffset.back().newOffset + it->sh_size;
+  }
+}
+
+// =====================================================================================================================
+// Rewrite ELF to add the new note section and its header for the cache hash.
+//
+// @param [in/out] elf      : ELF to be rewritten with the new .note section
+//                          for the cache hash.
+// @param offsetOfNewSectionHeader   : The offset where the new section header will be
+//                          placed.
+// @param newNoteSectionHeader      : The section header for the new note section.
+// @param zerosForAlignment : Zeros for the note section alignment between the
+//                          end of section header table and the section for
+//                          cache hash note.
+// @param noteSectionStream       : The note secion containing the cache hash.
+// @param sectionAndNewOffset  : The information to shift sections after the
+//                          section header table.
+void rewriteELFWithCacheHash(SmallVectorImpl<char> &elf, const ELF::Elf64_Off offsetOfNewSectionHeader,
+                             const ELF::Elf64_Shdr &newNoteSectionHeader, const SmallString<4> &zerosForAlignment,
+                             AppendingBinaryByteStream &noteSectionStream,
+                             SmallVectorImpl<SectionShiftInfo> &sectionAndNewOffset) {
+  // Strip sections after the section header table.
+  elf.resize(offsetOfNewSectionHeader);
+
+  // Increase the number of section headers in ELF header.
+  reinterpret_cast<ELF::Elf64_Ehdr *>(elf.data())->e_shnum++;
+
+  // Write the section header for the new note section.
+  raw_svector_ostream elfStream(elf);
+  elfStream << StringRef(reinterpret_cast<const char *>(&newNoteSectionHeader), sizeof(ELF::Elf64_Shdr));
+
+  // Write zeros for alignment before the new section.
+  elfStream << zerosForAlignment.str();
+
+  // Write the new note section.
+  ArrayRef<uint8_t> noteSection;
+  auto error = noteSectionStream.readBytes(0, noteSectionStream.getLength(), noteSection);
+  assert(!error);
+  elfStream << StringRef(reinterpret_cast<const char *>(noteSection.data()), noteSection.size());
+
+  // Sort SectionShiftInfo by the new offset of each section in the increasing
+  // order.
+  sort(sectionAndNewOffset,
+       [](const SectionShiftInfo &i0, const SectionShiftInfo &i1) { return i0.newOffset < i1.newOffset; });
+
+  // Shift the sections after section header table.
+  for (auto it = sectionAndNewOffset.begin(); it != sectionAndNewOffset.end(); ++it) {
+    elfStream.write_zeros(it->newOffset - elfStream.str().size());
+    elfStream << it->section.str();
+  }
+}
+
+} // anonymous namespace
+
+// -add-hash-to-elf
+static cl::opt<bool> AddHashToELF("add-hash-to-elf",
+                                  cl::desc("Add a .note section to ELF for hash used to lookup cache"),
+                                  cl::init(false));
```

Comment:
Can you move this to the top of the file? I think it's an unwritten rule to keep all options at the top.
You can put it inside the anonymous namespace, it there's no point to make it static when you have an existing anonymous namespace in place. The two should have identical semantics.

Reply:
done